requirements.txt

```diff
@@ -21,6 +21,6 @@ tensorboard>=1.15.0
 tqdm>=4.30
 scipy
 torchio>=0.18.68
-torch~=1.8.1
-torchvision~=0.9.1
+torch>=1.8.1,<=1.11.0
+torchvision>=0.9.1,<=0.12.0
```

Comment:
Removing the upper bound on the version exposes us to potential issues if some breaking changes versions are made in future versions of either packages, isn't it? I would only keep the versions that we did tested.

Reply:
Absolutely... In hindsight, we're already making a big transition and I guess it would be better if we have an upper bound to narrow down in case we find any unusual behaviour.

Reply:
Absolutely... In hindsight, we're already making a big transition and I guess it would be better if we have an upper bound to narrow down in case we find any unusual behaviour.

Reply:
But are you saying we should test every version of every dependency implied by our `>=min, <=max` bounds, @lifetheater57? I don't think there's python tooling that can do that, and even if there was the number of test cases would explode -- and our tests [already take half an hour per commit](https://github.com/ivadomed/ivadomed/actions/workflows/run_tests.yml).

I think the best we can do is set the `max` bound to the current latest `torch` (same with our other dependencies). If we ever miss updating our `max` for one of their releases we'll just have to assume it worked if the later one did. And if it doesn't we can go find the most recent version that did and pin to that.

Reply:
I ran into a real-world example of lacking an upper bound causing problems last week: https://github.com/zalandoresearch/pytorch-ts/issues/105

This project is [declares](https://github.com/zalandoresearch/pytorch-ts/blob/v0.6.0/setup.py#L19) `gluonts>=0.9.0`, sometime in the last few months gluonts [moved their API all around](https://github.com/awslabs/gluon-ts/issues/1575) without even writing a good changelog, so now

And now [`torchts==0.6.0`](https://pypi.org/project/pytorchts/0.6.0/#files) is broken _forevermore_. `pip install torchts==0.6.0` will always produce a broken install, so there's no way for people to go back in time and try an old version.

When I crashed into this I scrambled, and looked at the [release dates](https://pypi.org/project/pytorchts/#history) of [both projects](https://pypi.org/project/gluonts/#history) to figure out that `gluonts==0.9.3` _was_ the current version at the time of it's release, and I was able to get it working with:

```
pip install 'torchts' 'gluonts<0.9.4'
```

So we will be a lot friendlier to the future if we put `<max` bounds on every one of our dependencies.

SCT had the same problem, by the way, but they solved it by ignoring most of python's packaging conventions; instead, making a release means [generating and committing `requirements-freeze.txt`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/45b1c1759f991332697b4120160303e963f7a3c7/requirements-freeze.txt) to the `release` branch, and then the (homegrown, unconventional, non-`pip`) install script [will install all of them](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/0c17f870d82fea5518a875031b661375a920e404/install_sct#L629-L632). I don't think that's good either because it overspecifies the dependencies and requires all this fragile shell scripting to get it right, and then it's impossible for anyone to build on SCT because SCT controls the entire environment.

Reply:
On the other hand (and I'm thinking out loud here, trying to understand all this, so feel free to chime in and help us all understand), [Bernat's article](https://bernat.tech/posts/version-numbers/) makes the extremely good point that 

> [if two of your dependencies have conflicting dependency declarations of their own] Either way, your application no longer works. The only way to make it work is to either pin the service library down to the last working patch number, or upgrade your version pinning of tenacity. This is generating extra work for you with minimal benefit. Often it might not be even possible to use two conflicting libraries until one of them relaxes their requirements.

and

> Given the scarce human resources to maintain a library, in practice, there’s a single supported version for any library at any given point in time: the latest one. Any version before that (be that major, minor, patch) is in essence abandoned:
>
> *    if you want security updates, you need to move to the latest version,
> *    if you wish to a bugfix you need to move to the newest version,
> *    if you want a new feature, it is only going to be available in the latest version.
>
> You get the idea. Unless you’re thrilled with your current version, to pull in any change you will need to move to the last released version (be that security improvement, bugfix, or feature). Otherwise, you’ll not get it.


Essentially he's advocating using rolling releases -- where the *only* version that is promised to work is "right now" (like ArchLinux, Alpine, or Kali Linux) -- instead of point releases with multiple branches and backports like Ubuntu/Debian/CentOS/OpenBSD.

If you look through https://packages.archlinux.org, for example, say, [here](https://github.com/archlinux/svntogit-packages/blob/d8a74408dfd8390a542a800b10d8f35f6d46e6b4/trunk/PKGBUILD#L11), you'll see that those packages almost never pin dependencies. They do [support it](https://man.archlinux.org/man/PKGBUILD.5#OPTIONS_AND_DIRECTIVES)

>  Entries can also include a version requirement of the form name<>version, where <> is one of five comparisons: >= (greater than or equal to), <= (less than or equal to), = (equal to), > (greater than), or < (less than). 

but I think because Arch is a rolling release and the _only_ support version is "right now", then there is rarely a reason to use version pinning, it's easier to just fix the actual bug. Personally I love this as a user, it means things get fixed faster, and I always get the latest software. Debian and Ubuntu are always months, sometimes years behind -- and I find myself leaning more and more on `conda` when I have to work on Debian or Ubuntu. I've heard that Debian Testing and OpenBSD-current are also very nice to use and are what most people "in the know" use -- but they should be, they are also a rolling release.


I think, given that we are not Canonical Inc. with a huge team that can be spent on tracking the minutae of version updates, and we aren't maintaining multiple release branches anyway we should consider adopting rolling releases, and **unpinning everything**.

We can keep `torch>=1.8.0,torch<1.9.0` maybe since we know we're not yet compatible with other `torch`es, but otherwise do we really need to pin anything?

This will mean the **downside** that someone trying to reproduce results with an old version by `pip install ivadomed==2.2.0` won't be able to, but the **upside** that we make ourselves friendlier to being integrated with other projects (by not overconstraining).

Old ivadomeds already don't work right anyway. The most obvious example I know is at the moment `pip install ivadomed` on python 3.10, it ends up getting `ivadomed==2.3.1` (because that's the most recent version that _did not_ declare `python_requires` and is compatible with all the declared constraints), but it ends up getting the latest torch 1.12 and onnxruntime 1.12 which I am guessing it is not compatible with. Maybe we should just embrace this.

<details>

```
p115628@joplin:~$ pip install ivadomed
Defaulting to user installation because normal site-packages is not writeable
Collecting ivadomed
  Downloading ivadomed-2.6.1-py3-none-any.whl (145 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 145.2/145.2 kB 5.4 MB/s eta 0:00:00
Collecting pandas
  Downloading pandas-1.4.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (11.6 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.6/11.6 MB 60.8 MB/s eta 0:00:00
Collecting tensorboard
  Using cached tensorboard-2.9.1-py3-none-any.whl (5.8 MB)
Collecting ivadomed
  Downloading ivadomed-2.6.0-py3-none-any.whl (139 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 139.1/139.1 kB 30.7 MB/s eta 0:00:00
  Downloading ivadomed-2.5.0-py3-none-any.whl (137 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 137.2/137.2 kB 51.0 MB/s eta 0:00:00
  Downloading ivadomed-2.4.0-py3-none-any.whl (132 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 132.8/132.8 kB 31.0 MB/s eta 0:00:00
Requirement already satisfied: h5py in /usr/local/lib/python3.10/dist-packages (from ivadomed) (3.7.0)
Collecting onnxruntime
  Downloading onnxruntime-1.12.1-cp310-cp310-manylinux_2_27_x86_64.whl (4.9 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.9/4.9 MB 69.4 MB/s eta 0:00:00
Requirement already satisfied: pillow>=7.0.0 in /usr/lib/python3/dist-packages (from ivadomed) (9.0.1)
Collecting pytest
  Using cached pytest-7.1.2-py3-none-any.whl (297 kB)
Collecting scikit-image
  Downloading scikit_image-0.19.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (13.9 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.9/13.9 MB 61.7 MB/s eta 0:00:00
Collecting ivadomed
  Downloading ivadomed-2.3.1-py3-none-any.whl (129 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 129.2/129.2 kB 27.9 MB/s eta 0:00:00
Requirement already satisfied: torch>=1.5.0 in ./.local/lib/python3.10/site-packages (from ivadomed) (1.11.0)
Collecting torchvision>=0.6.0
  Downloading torchvision-0.13.0-cp310-cp310-manylinux1_x86_64.whl (19.1 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.1/19.1 MB 49.3 MB/s eta 0:00:00
Requirement already satisfied: tqdm>=4.30 in /usr/lib/python3/dist-packages (from ivadomed) (4.57.0)
Requirement already satisfied: joblib in ./.local/lib/python3.10/site-packages (from ivadomed) (1.1.0)
Requirement already satisfied: nibabel in ./.local/lib/python3.10/site-packages (from ivadomed) (3.2.2)
Collecting scikit-learn>=0.20.3
  Downloading scikit_learn-1.1.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (30.5 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 30.5/30.5 MB 39.6 MB/s eta 0:00:00
Collecting seaborn
  Using cached seaborn-0.11.2-py3-none-any.whl (292 kB)
Collecting bids-neuropoly>=0.2
  Downloading bids_neuropoly-0.2-py3-none-any.whl (3.1 kB)
Requirement already satisfied: matplotlib>=3.3.0 in /usr/local/lib/python3.10/dist-packages (from ivadomed) (3.5.2)
Collecting pybids>=0.6.5
  Downloading pybids-0.15.2-py3-none-any.whl (14.9 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.9/14.9 MB 62.6 MB/s eta 0:00:00
Requirement already satisfied: numpy>=1.17 in /usr/lib/python3/dist-packages (from matplotlib>=3.3.0->ivadomed) (1.21.5)
Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.3.0->ivadomed) (1.4.3)
Requirement already satisfied: pyparsing>=2.2.1 in /usr/lib/python3/dist-packages (from matplotlib>=3.3.0->ivadomed) (2.4.7)
Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.3.0->ivadomed) (4.33.3)
Requirement already satisfied: packaging>=20.0 in ./.local/lib/python3.10/site-packages (from matplotlib>=3.3.0->ivadomed) (21.3)
Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.10/dist-packages (from matplotlib>=3.3.0->ivadomed) (0.11.0)
Requirement already satisfied: python-dateutil>=2.7 in ./.local/lib/python3.10/site-packages (from matplotlib>=3.3.0->ivadomed) (2.8.2)
Requirement already satisfied: setuptools in ./.local/lib/python3.10/site-packages (from nibabel->ivadomed) (62.5.0)
Requirement already satisfied: pytz>=2020.1 in ./.local/lib/python3.10/site-packages (from pandas->ivadomed) (2022.1)
Requirement already satisfied: scipy>=1.3.2 in /usr/lib/python3/dist-packages (from scikit-learn>=0.20.3->ivadomed) (1.8.0)
Collecting threadpoolctl>=2.0.0
  Using cached threadpoolctl-3.1.0-py3-none-any.whl (14 kB)
Requirement already satisfied: typing-extensions in ./.local/lib/python3.10/site-packages (from torch>=1.5.0->ivadomed) (4.2.0)
Requirement already satisfied: requests in /usr/local/lib/python3.10/dist-packages (from torchvision>=0.6.0->ivadomed) (2.28.0)
Collecting torch>=1.5.0
  Downloading torch-1.12.0-cp310-cp310-manylinux1_x86_64.whl (776.3 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 776.3/776.3 MB 2.3 MB/s eta 0:00:00
Requirement already satisfied: sympy in ./.local/lib/python3.10/site-packages (from onnxruntime->ivadomed) (1.10.1)
Requirement already satisfied: protobuf in ./.local/lib/python3.10/site-packages (from onnxruntime->ivadomed) (3.20.1)
Requirement already satisfied: flatbuffers in ./.local/lib/python3.10/site-packages (from onnxruntime->ivadomed) (2.0)
Collecting coloredlogs
  Using cached coloredlogs-15.0.1-py2.py3-none-any.whl (46 kB)
Requirement already satisfied: tomli>=1.0.0 in ./.local/lib/python3.10/site-packages (from pytest->ivadomed) (2.0.1)
Collecting pluggy<2.0,>=0.12
  Using cached pluggy-1.0.0-py2.py3-none-any.whl (13 kB)
Collecting iniconfig
  Using cached iniconfig-1.1.1-py2.py3-none-any.whl (5.0 kB)
Requirement already satisfied: attrs>=19.2.0 in ./.local/lib/python3.10/site-packages (from pytest->ivadomed) (21.4.0)
Collecting py>=1.8.2
  Using cached py-1.11.0-py2.py3-none-any.whl (98 kB)
Collecting tifffile>=2019.7.26
  Downloading tifffile-2022.8.3-py3-none-any.whl (207 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.9/207.9 kB 38.9 MB/s eta 0:00:00
Collecting PyWavelets>=1.1.1
  Downloading PyWavelets-1.3.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (6.9 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.9/6.9 MB 85.4 MB/s eta 0:00:00
Collecting networkx>=2.2
  Using cached networkx-2.8.5-py3-none-any.whl (2.0 MB)
Collecting imageio>=2.4.1
  Downloading imageio-2.21.0-py3-none-any.whl (3.4 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.4/3.4 MB 82.4 MB/s eta 0:00:00
Requirement already satisfied: google-auth<3,>=1.6.3 in ./.local/lib/python3.10/site-packages (from tensorboard->ivadomed) (2.6.6)
Requirement already satisfied: markdown>=2.6.8 in /usr/lib/python3/dist-packages (from tensorboard->ivadomed) (3.3.6)
Collecting absl-py>=0.4
  Downloading absl_py-1.2.0-py3-none-any.whl (123 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 123.4/123.4 kB 43.4 MB/s eta 0:00:00
Collecting tensorboard-data-server<0.7.0,>=0.6.0
  Using cached tensorboard_data_server-0.6.1-py3-none-manylinux2010_x86_64.whl (4.9 MB)
Collecting tensorboard-plugin-wit>=1.6.0
  Using cached tensorboard_plugin_wit-1.8.1-py3-none-any.whl (781 kB)
Collecting grpcio>=1.24.3
  Downloading grpcio-1.47.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.5 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.5/4.5 MB 78.2 MB/s eta 0:00:00
Collecting protobuf
  Downloading protobuf-3.19.4-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.1 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 87.5 MB/s eta 0:00:00
Collecting google-auth-oauthlib<0.5,>=0.4.1
  Using cached google_auth_oauthlib-0.4.6-py2.py3-none-any.whl (18 kB)
Requirement already satisfied: wheel>=0.26 in /usr/lib/python3/dist-packages (from tensorboard->ivadomed) (0.37.1)
Requirement already satisfied: werkzeug>=1.0.1 in ./.local/lib/python3.10/site-packages (from tensorboard->ivadomed) (2.1.2)
Requirement already satisfied: cachetools<6.0,>=2.0.0 in ./.local/lib/python3.10/site-packages (from google-auth<3,>=1.6.3->tensorboard->ivadomed) (5.2.0)
Requirement already satisfied: rsa<5,>=3.1.4 in /usr/lib/python3/dist-packages (from google-auth<3,>=1.6.3->tensorboard->ivadomed) (4.8)
Requirement already satisfied: pyasn1-modules>=0.2.1 in ./.local/lib/python3.10/site-packages (from google-auth<3,>=1.6.3->tensorboard->ivadomed) (0.2.8)
Requirement already satisfied: six>=1.9.0 in /usr/lib/python3/dist-packages (from google-auth<3,>=1.6.3->tensorboard->ivadomed) (1.16.0)
Requirement already satisfied: requests-oauthlib>=0.7.0 in ./.local/lib/python3.10/site-packages (from google-auth-oauthlib<0.5,>=0.4.1->tensorboard->ivadomed) (1.3.1)
Collecting sqlalchemy<1.4.0.dev0
  Downloading SQLAlchemy-1.3.24.tar.gz (6.4 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.4/6.4 MB 82.6 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
Collecting formulaic<0.4,>=0.2.4
  Using cached formulaic-0.3.4-py3-none-any.whl (68 kB)
Collecting bids-validator
  Downloading bids_validator-1.9.7-py2.py3-none-any.whl (21 kB)
Requirement already satisfied: click>=8.0 in ./.local/lib/python3.10/site-packages (from pybids>=0.6.5->bids-neuropoly>=0.2->ivadomed) (8.1.3)
Collecting num2words
  Downloading num2words-0.5.11-py3-none-any.whl (116 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 36.9 MB/s eta 0:00:00
```

</details>

Reply:
Another very different approach I've seen is to allow any version of a package to be installed side-by-side. For example [distr1](https://michael.stapelberg.ch/posts/2019-08-17-introducing-distri/#key-idea-exchange-directories) and [nix](https://nixos.org/).  Packages get installed to `/path/to/dependency-$DEP_VERSION` and then any user of it like `/path/to/app-$APP_VERSION` is arranged so that the environment it sees prefers loading libraries from `/path/to/dependency-$DEP_VERSION` (e.g. in python you can do this by prepending to `sys.path` or using .pth files)

And sometimes even more traditional package systems use this too, for example the long period when most Linuces had a `python2` and a `python3` package, side by side. In that case Ubuntu et al. just moved part of the version number from the version field into the name field. That worked, though it meant every python library they distributed had to come in two versions: `python2-requests`, `python3-requests` etc. NixOS's version standardizes and enhances this.

Reply:
tl;dr:

We can?

* **rolling releases**: **remove** all dependency specifiers entirely; document that the _only_ ivadomed we expect to work is "right now"
  * automate: use a cronjob to test `master` against all the _latest_ dependencies every day 
* **pinned releases**: put `<max` bounds on **all** our dependencies
  * automate: rely on [Dependabot](https://github.com/dependabot) to remind us bump + test when new releases come out

Reply:
@kousu  
> I think, given that we are not Canonical Inc. with a huge team that can be spent on tracking the minutae of version updates, and we aren't maintaining multiple release branches anyway we should consider adopting rolling releases, and unpinning everything.

At the same time, given that we are not a big team we cannot test new releases of our dependencies as soon as they get out (I which we could automate the testing of everything!).

> I think because Arch is a rolling release and the only support version is "right now", then there is rarely a reason to use version pinning, it's easier to just fix the actual bug. Personally I love this as a user, it means things get fixed faster, and I always get the latest software. Debian and Ubuntu are always months, sometimes years behind

The downside of the Arch way is the need to have pro-active expert users that can deal with the problems when something breaks.

> But are you saying we should test every version of every dependency implied by our >=min, <=max bounds, @lifetheater57 ?

No, I would only test the `min` and `max` versions of the libraries we "manually" depends on if it is possible to automate that. For example, I would do it for `PyTorch`, but not for a library that is automatically installed by `PyTorch`.

Reply:
> tl;dr:
> 
> We can?
> 
>     * **rolling releases**: entirely remove dependency specifiers
>       
>       * **document**: the _only_ ivadomed we expect to work is "right now"
>       * **automate**: use a cronjob to test `master` against all the _latest_ dependencies every day
> 
>     * **pinned releases**: put `<max` bounds on **all** our dependencies
>       
>       * **automate**: rely on [Dependabot](https://github.com/dependabot) to remind us bump + test when new releases come out

At first, I would have prefered pinning the versions. Increasing the `max` version when we can test the newest one. Increasing the `min` when we either need a fix or a feature available in a newer version.

However, the more I think about rolling releases, the more I like the idea. What are the major downsides of going the "rolling releases" way?

Reply:
>  No, I would only test the min and max versions of the libraries we "manually" depends on if it is possible to automate that. For example, I would do it for PyTorch, but not for a library that is automatically installed by PyTorch.

Excellent, I agree.

> What are the major downsides of going the "rolling releases" way?

I'm not totally sure. So far I've thought up two: it would leave us open to your [original concern](https://github.com/ivadomed/ivadomed/pull/1137#discussion_r938281300) of surprise breakage, and it would mean that past versions will be broken forever, instead of just what's on the `master` branch.

I think for the sake of _this_ PR, I am happy to leave the upper bound. We can keep thinking about this and decide the best practice for all our dependencies later.

I think this PR is probably ready to merge.

Reply:
> > What are the major downsides of going the "rolling releases" way?
> 
> I'm not totally sure. So far I've thought up two: it would leave us open to your [original concern](https://github.com/ivadomed/ivadomed/pull/1137#discussion_r938281300) of surprise breakage

In practice the "surprise breakage" of rolling releases means no one can install a working ivadomed, so we need to be able to issue hotfixes rapidly.

On the other hand, pinning dependencies doesn't avoid breakage, it just makes it someone else's problem: anyone who is trying to combine ivadomed with, say, a pytorch script they wrote that needs the very latest torch will be stuck. Though they can work around it with a virtualenv.

Reply:
> tl;dr:
> 
> We can?
> 
> * **rolling releases**: entirely remove dependency specifiers
>   
>   * **document**: the _only_ ivadomed we expect to work is "right now"
>   * **automate**: use a cronjob to test `master` against all the _latest_ dependencies every day
> * **pinned releases**: put `<max` bounds on **all** our dependencies
>   
>   * **automate**: rely on [Dependabot](https://github.com/dependabot) to remind us bump + test when new releases come out

I pondered a bit upon this and I don't feel we're there yet to go with rolling releases, especially thinking about the integration of Multi-Contrast Multi-Session. I'm quite content with pinned releases as of now.

Reply:
Thanks for all your thoughts. I'm going to start a new issue to keep talking about all this.

Reply:
-> https://github.com/ivadomed/ivadomed/issues/1191